src/set_theory/fixed_points.lean

```diff
@@ -0,0 +1,316 @@
+three forms: as statements about type-indexed families of normal functions, as statements about
+ordinal-indexed families of normal functions, and as statements about a single normal function. For
+the most part, the first case encompasses the others.
+
+Moreover, we prove some lemmas about the fixed points of specific normal functions.
+
+## Main definitions and results
+
+* `nfp_family`, `nfp_bfamily`, `nfp`: the next fixed point of a (family of) normal function(s).
+* `fp_family_unbounded`, `fp_bfamily_unbounded`, `fp_unbounded`: the (common) fixed points of a
+  (family of) normal function(s) are unbounded in the ordinals.
+* `deriv_add_eq_mul_omega_add`: a characterization of the derivative of addition.
+-/
+
+noncomputable theory
+
+universes u v
+
+open function
+
+namespace ordinal
+
+/-! ### Fixed points of type-indexed families of ordinals -/
+
+section
+variables {ι : Type u} {f : ι → ordinal.{max u v} → ordinal.{max u v}}
+
+/-- The next common fixed point, at least `a`, for a family of normal functions. -/
+def nfp_family (f : ι → ordinal → ordinal) (a) : ordinal :=
+sup (list.foldr f a)
+
+theorem nfp_family_eq_sup (f : ι → ordinal → ordinal) (a) :
+  nfp_family f a = sup (list.foldr f a) :=
+rfl
+
+theorem foldr_le_nfp_family (f : ι → ordinal → ordinal) (a l) :
+  list.foldr f a l ≤ nfp_family f a :=
+le_sup _ _
+
+theorem self_le_nfp_family (f : ι → ordinal → ordinal) (a) : a ≤ nfp_family f a :=
+le_sup _ []
+
+theorem lt_nfp_family {a b} : a < nfp_family f b ↔ ∃ l, a < list.foldr f b l :=
+lt_sup
+
+theorem nfp_family_le {a b} : nfp_family f a ≤ b ↔ ∀ l, list.foldr f a l ≤ b :=
+sup_le
+
+theorem nfp_family_monotone (hf : ∀ i, monotone (f i)) : monotone (nfp_family f) :=
+λ a b h, sup_le.2 $ λ l, (list.foldr_monotone hf l h).trans (le_sup _ l)
+
+theorem apply_lt_nfp_family (H : ∀ i, is_normal (f i)) {a b} (hb : b < nfp_family f a) (i) :
+  f i b < nfp_family f a :=
+let ⟨l, hl⟩ := lt_nfp_family.1 hb in lt_sup.2 ⟨i :: l, (H i).strict_mono hl⟩
+
+theorem apply_lt_nfp_family_iff [nonempty ι] (H : ∀ i, is_normal (f i)) {a b} :
+  (∀ i, f i b < nfp_family f a) ↔ b < nfp_family f a :=
+⟨λ h, lt_nfp_family.2 $ let ⟨l, hl⟩ := lt_sup.1 (h (classical.arbitrary ι)) in
+  ⟨l, ((H _).self_le b).trans_lt hl⟩, apply_lt_nfp_family H⟩
+
+theorem nfp_family_le_apply [nonempty ι] (H : ∀ i, is_normal (f i)) {a b} :
+  (∃ i, nfp_family f a ≤ f i b) ↔ nfp_family f a ≤ b :=
+by { rw ←not_iff_not, push_neg, exact apply_lt_nfp_family_iff H }
+
+theorem nfp_family_le_fp (H : ∀ i, monotone (f i)) {a b} (ab : a ≤ b) (h : ∀ i, f i b ≤ b) :
+  nfp_family f a ≤ b :=
+sup_le.2 $ λ l, begin
+  by_cases hι : is_empty ι,
+  { rwa @unique.eq_default _ (@list.unique_of_is_empty ι hι) l },
+  { haveI := not_is_empty_iff.1 hι,
+    induction l with i l IH generalizing a, {exact ab},
+    exact (H i (IH ab)).trans (h i) }
+end
+
+theorem nfp_family_fp (H : ∀ i, is_normal (f i)) (i a) : f i (nfp_family f a) = nfp_family f a :=
```

Comment:
Pretty minor but I think you can change `H` to just say that `f i` is normal, rather than for every `i`

Reply:
I actually can't. Consider the case when `f` is a family of two functions, consisting of `f 0 = id` and `f 1 = λ x, x + 1`. Then `nfp_family f 0 = ω` but `f 1 (nfp_family f 0) = ω + 1`.

Reply:
I don't follow.

Reply:
The proof
```
theorem nfp_family_fp (i a) (H : is_normal (f i)) : f i (nfp_family f a) = nfp_family f a :=
begin
  unfold nfp_family,
  rw @is_normal.sup _ H _ _ ⟨[]⟩,
  apply le_antisymm;
  refine ordinal.sup_le (λ l, _),
  { exact le_sup _ (i :: l) },
  { exact (H.self_le _).trans (le_sup _ _) }
end
```
works just as well.

Reply:
Further, this lets you weaken the hypotheses in many of the following lemmas

Reply:
My bad then. I've applied this change and weakened the corresponding hypotheses.